Updated task PUT

Vatsal09 opened the pull request:
Updated PUT to stop erroring out when one task fails to update, from a list of tasks

__What__

- [x] Updated the PUT logic
- [x] Streamlined the conditionals
- [x] Added exception logging when any task fails to update due to a DB query error
- [x] Removed a lot of comments and formatted file

__Why__
In order to enable task update from the app, including complete task

__How__

- [x]  unit tests
- [x]  functional tests
- [x] regression tests

joshpaulchan commented on resources/Task.py:
```diff
@@ -237,11 +237,14 @@ def post(self):
-            return error_response("POST payload missing")
+            return error_response("Empty POST request body")
 
         if ("userId" not in json_data) or (json_data["userId"] is None):
```
guess we missed this the first time around but `if not json_data.get("userId")` should be fine here

joshpaulchan commented on resources/Task.py:
```diff
@@ -292,93 +295,109 @@ def put(self):
 
-                    for user_id in get_user_to_task_ids_to_update(
-                        task_data.get("taskId"), task_data["add_users"]
+                    if ("add_users" in task_data.keys()) and (
```
Same thing here: `if task_date.get("add_users"):`

joshpaulchan replied:
Algorithmically speaking (and potentially performance-wise), what do you think about accumulating these relation changes and doing them all at once? i.e. right now we're basically doing:

```python
for task_change in task_changes:
    task = get_task(task_change)
    update_task(task, task_change)

    add_users(task, task_change.users)
    remove_users(task, task_change.users)
    
    add_to_tasklist(task, task_change.lists)
    remove_from_tasklist(task, task_change.lists)
```

What do you think of doing something like:

```python
users_to_add_for_tasks = {} # task -> user[]
users_to_remove_for_tasks = {} # task -> user[]

lists_to_add_for_tasks = {} # task -> tasklists[]
lists_to_remove_for_tasks = {} # task -> tasklists[]
for task_change in task_changes:
    task = get_task(task_change)
    update_task(task, task_change)

    users_to_add_for_task[task.id] = task_changes.users_to_add
    users_to_remove_for_task[task.id] = task_changes.users_to_remove

    lists_to_add_for_task[task.id] = task_changes.users_to_add
    lists_to_remove_for_task[task.id] = task_changes.users_to_remove

# very roughly speaking, get all the user associations for the tasks we're changing in one go (should equate to 1 SELECT)
users_for_tasks = get_users_for_tasks(users_to_add.taskids + users_to_add.taskIds)

# also very rough, add the users-task associations we want to add then remove the ones we don't want (translates to some deletes + some inserts)
users_for_tasks + users_to_add - users_to_remove

# then commit the transaction and save it
users_for_tasks.save()

# same thing for task-tasklist associations pretty much
```

I can go into more detail if things are unclear but wanted to run this by you and see what you think.

joshpaulchan commented on resources/Task.py:
```diff
@@ -292,93 +295,109 @@ def put(self):
 
-                    for tasklist_id in get_task_to_tasklist_ids_to_update(
-                        task_data.get("taskId"), task_data["add_tasklists"]
+                    if ("add_tasklists" in task_data.keys()) and (
```
ditto

joshpaulchan commented on resources/Task.py:
```diff
@@ -292,93 +295,109 @@ def put(self):
 
-                    for user_id in get_users_to_del(
-                        task_data.get("taskId"), task_data["del_users"]
+                    if ("del_users" in task_data.keys()) and (
```
ditto

joshpaulchan commented on resources/Task.py:
```diff
@@ -292,93 +295,109 @@ def put(self):
 
-                    for tasklist in get_tasklists_to_del(
-                        task_data.get("taskId"), task_data["del_tasklists"]
+                    if ("del_tasklists" in task_data.keys()) and (
```
ditto

joshpaulchan commented on resources/Task.py:
```diff
@@ -292,93 +295,109 @@ def put(self):
+                    return_response_errors.append(
+                        {
+                            "taskId": task_data.get("taskId"),
+                            "error": "Exception error. Check logs",
```
Probably better to say there was an unexpected exception there, we already know to check the logs

joshpaulchan commented on resources/Task.py:
```diff
@@ -292,93 +295,109 @@ def put(self):
-        result = tasks_schema.dump(get_tasks_by_ids(task_ids_bulk_get)).data
-        return success_response(result)
+        result = tasks_schema.dump(get_tasks_by_ids(bulk_get_tasks_by_ids)).data
+        return {"data": result, "errors": return_response_errors}, 200
```
NP: maybe use `success( ... )` helper here?

Vatsal09 commented:
@joshpaulchan you can check now. Big changes: rehauled PUT, deleted obsolete methods, cleaned up remaining methods, moved helper methods to its own file, added logging for helper methods that access db, and removed all associations of Tasklist. I don't think it would be wise to keep Tasklist since we aren't using it and we are planning to add Events to the same code base

joshpaulchan commented on resources/Task.py:
```diff
@@ -237,154 +59,143 @@ def post(self):
-                print(e)
-                return error_response(
-                    "One or more taskId(s) does not exist in the database"
+            if not task_data.get("taskId"):
```
just curious: could we define a new marshmallow schema for `task_change` (or something) with this field and embed the task-level `task_schema` such that this case is handled by schema validator?

Vatsal09 replied:
We can make fields required but then we have to define multiple schemas. For example, we can't make task_id a requirement if we plan to use the schema in POST, where the user won't know the id before the task has been created. For readability and consistency, I reckoned it would be best to use a singular Schema validation. However, I have updated the schema to optimize for PUT method. Now schema validation checks for task_id

joshpaulchan commented on resources/Task.py:
```diff
@@ -237,154 +59,143 @@ def post(self):
+        bulk_get_tasks_by_ids = []
+        # Collect user & task ids for a single UserToTask SELECT
+        bulk_get_users_user_ids = []
+        bulk_get_users_task_ids = []
```
Q: Would these be better suited as sets?

Vatsal09 replied:
Makes sense

Vatsal09 replied:
Made changes

joshpaulchan commented on resources/Task.py:
```diff
@@ -237,154 +59,143 @@ def post(self):
+                )
+            else:
+                add_users = (
+                    set(task_data.get("addUsers"))
```
NP: we can probably do: `users_to_add = set(task_data.get("addUsers", []))`

Vatsal09 replied:
Done

joshpaulchan commented on resources/Task.py:
```diff
@@ -237,154 +59,143 @@ def post(self):
+        # Collect errors for any task that is not processed
+        return_response_errors = []
+        # Collect task ids for a single Task SELECT
+        bulk_get_tasks_by_ids = []
```
Suggestion: This is unnecessary since `tasks_to_update` is keyed by `task.id` and de-duplicated as well, so we just fetch all affected tasks using: `get_tasks...(tasks_to_update.keys())`. I think we can remove this.

joshpaulchan replied:
Looking a bit more, `bulk_get_users_task_ids` is also unnecessary due to this (given how you fetch them with an AND clause), which should fetch the subset you want.

Vatsal09 replied:
Agree with using just tasks_to_update.keys(). Need the bulk_get_users_task_ids collection because we could get a change like [{taskId:1, title: "eat"}, {taskId:2, addUsers: [56,78], deleteUsers: [5,6]}] and if we query using all the taskIds we will get UserToTasks like 1-56, 1-78, 1-5, 1-6. Since the logic checks if the keys addUsers and deleteUsers are present, it will not make the change but there is all this overhead in memory and looping to create the defaultdicts

joshpaulchan replied:
oh yah that makes sense. We may get extra `UserToTask`s we are not operating on. good thinking!

joshpaulchan commented on resources/Task.py:
```diff
@@ -237,154 +59,143 @@ def post(self):
+                del_user_to_tasks = []
+                for del_user_to_task_id in (task_changes.get("del_users")).intersection(grouped_task_id_to_user_ids.get(task_id, set())):
+                    del_user_to_tasks.append(grouped_user_to_tasks[task_id,del_user_to_task_id])
+                for user_to_task in del_user_to_tasks:
```
Hm, this following aggregation and iteration seems unnecessary. Can we not just call `db.session.delete()` on the `grouped_user_to_tasks[task_id, del_user_to_task_id]` instead of building up a new list?

Vatsal09 replied:
Yeah not sure what I was thinking. We can mark the UserToTasks for deletion right in the 1st for loop

joshpaulchan commented on resources/Task.py:
```diff
@@ -237,154 +59,143 @@ def post(self):
+            tasks_to_bulk_update.append(task)
+
+        db.session.add_all(tasks_to_bulk_update)
+        db.session.commit()
```
Since this a complex + critical path (tasks will probably be updated 100x more than they will be created/deleted) can we add some unit tests for this? I have some ideas about how to go about doing that if you'd like to pair on it.

joshpaulchan commented on resources/task_helper_functions.py:
```diff
@@ -0,0 +1,165 @@
+# task_helper_functions.py contain helper functions used by task.py
```
Suggestion: What do u think about naming this file `repository` or `dao` since the majority of these are persistence functions? that way we know this file is just for persistence-related logic. (would also move the 2 api-related functions to an `api_util.py` haha)

Vatsal09 replied:
Done

Vatsal09 commented on app.py:
```diff
@@ -1,14 +1,12 @@
-from resources.Task import TaskResource
-from resources.Tasklist import TasklistResource
+
+from resources.task import TaskResource
```
I have it locally changed to task.py instead of Task.py. git mv fixed this

Vatsal09 commented on resources/Task.py:
```diff
@@ -237,154 +59,143 @@ def post(self):
-            task.tasklists.append(tasklist)
-
-        # Finally commit all changes to database
+        task.users.append(new_user_to_task(task.id, json_data.get("userId")))
```
We were already validating the existence of userId in line 70 but now I have updated Marshmallow schema to check if userId and taskId are integers during load

Vatsal09 replied:
Verified it works using Postman
